File: 구현/boj_20056.py
import sys
input = sys.stdin.readline

dx = [-1, -1, 0, 1, 1, 1, 0, -1]
dy = [0, 1, 1, 1, 0, -1, -1, -1]

def move_fireballs(board):
    new_board = [[[] for __ in range(N)] for _ in range(N)]

    # 이동시키기
    for x in range(N):
        for y in range(N):
            if len(board[x][y]) == 0:
                continue
            
            for fb in board[x][y]:
                nx = (x + dx[fb[2]]*fb[1]) % N
                ny = (y + dy[fb[2]]*fb[1]) % N
                new_board[nx][ny].append(fb)

    # 합칠 거 합치기
    for x in range(N):
        for y in range(N):
            if len(new_board[x][y]) < 2:
                continue

            m_sum = sum([fb[0] for fb in new_board[x][y]])
            new_m = m_sum // 5

            if new_m == 0:
                new_board[x][y] = []
                continue
            
            s_sum = sum([fb[1] for fb in new_board[x][y]])
            new_s = s_sum // len(new_board[x][y])

            all_even = False
            if sum([1 for fb in new_board[x][y] if fb[2] % 2 == 0]) == len(new_board[x][y]):
                all_even = True

            all_odd = False
            if sum([1 for fb in new_board[x][y] if fb[2] % 2 != 0]) == len(new_board[x][y]):
                all_odd = True

            new_board[x][y] = []
            if all_even or all_odd:
                for new_d in [0, 2, 4, 6]:
                    new_board[x][y].append([new_m, new_s, new_d])
            else:
                for new_d in [1, 3, 5, 7]:
                    new_board[x][y].append([new_m, new_s, new_d])

    # 보드 갱신
    return new_board


def count_fireballs(board):
    count = 0;
    for x in range(N):
        for y in range(N):
            if len(board[x][y]) == 0:
                continue

            for fb in board[x][y]:
                count += fb[0]

    return count


N, M, K = map(int, input().split())
board = [[[] for __ in range(N)] for _ in range(N)]
for _ in range(M):
    r, c, m, s, d = map(int, input().split())
    board[r-1][c-1].append([m, s, d])

for _ in range(K): 
    board = move_fireballs(board)

print(count_fireballs(board))

File: 구현/boj_20056.java
package 구현;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

public class boj_20056 {
    static final int[] dx = {-1, -1, 0, 1, 1, 1, 0, -1};
    static final int[] dy = {0, 1, 1, 1, 0, -1, -1, -1};
    static int N, M, K;

    public static void main(String args[]) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        K = Integer.parseInt(st.nextToken());

        List<List<ArrayList<Fireball>>> board = new ArrayList<>();
        for (int i = 0; i < N; i++) {
            List<ArrayList<Fireball>> row = new ArrayList<>();
            for (int j = 0; j < N; j++) {
                row.add(new ArrayList<>());
            }
            board.add(row);
        }

        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            int x = Integer.parseInt(st.nextToken());
            int y = Integer.parseInt(st.nextToken());
            int m = Integer.parseInt(st.nextToken());
            int s = Integer.parseInt(st.nextToken());
            int d = Integer.parseInt(st.nextToken());
            board.get(x-1).get(y-1).add(new Fireball(m, s, d));
        }

        for (int i = 0; i < K; i++) {
            board = moveFireballs(board);
        }

        System.out.println(getRemainedFireballs(board));
    }

    static List<List<ArrayList<Fireball>>> moveFireballs(List<List<ArrayList<Fireball>>> board) {
        List<List<ArrayList<Fireball>>> newBoard = new ArrayList<>();
        for (int i = 0; i < N; i++) {
            List<ArrayList<Fireball>> row = new ArrayList<>();
            for (int j = 0; j < N; j++) {
                row.add(new ArrayList<>());
            }
            newBoard.add(row);
        }

        for (int x = 0; x < N; x++) {
            for (int y = 0; y < N; y++) {
                if (board.get(x).get(y).size() == 0) {
                    continue;
                }

                for (Fireball fb : board.get(x).get(y)) {
                    int nx = (x + dx[fb.d]*fb.s) % N;
                    int ny = (y + dy[fb.d]*fb.s) % N;
                    while (nx < 0) nx += N;
                    while (ny < 0) ny += N;
                    newBoard.get(nx).get(ny).add(new Fireball(fb.m, fb.s, fb.d));
                }
            }
        }

        for (int x = 0; x < N; x++) {
            for (int y = 0; y < N; y++) {
                if (newBoard.get(x).get(y).size() < 2) {
                    continue;
                }

                int mSum = 0, sSum = 0;
                boolean allEven = true, allOdd = true;
                for (Fireball fb : newBoard.get(x).get(y)) {
                    mSum += fb.m;
                    sSum += fb.s;

                    if (allEven && fb.d % 2 != 0) {
                        allEven = false;
                    }

                    if (allOdd && fb.d % 2 != 1) {
                        allOdd = false;
                    }
                }

                int newM = mSum / 5;
                int newS = sSum / newBoard.get(x).get(y).size();
                newBoard.get(x).get(y).clear();

                if (newM == 0) {
                    continue;
                }

                if (allOdd || allEven) {
                    for (int newD = 0; newD < 8; newD += 2) {
                        newBoard.get(x).get(y).add(new Fireball(newM, newS, newD));
                    }
                } else {
                    for (int newD = 1; newD < 8; newD += 2) {
                        newBoard.get(x).get(y).add(new Fireball(newM, newS, newD));
                    }
                }
            }
        }

        return newBoard;
    }

    static int getRemainedFireballs(List<List<ArrayList<Fireball>>> board) {
        int count = 0;

        for (int x = 0; x < N; x++) {
            for (int y = 0; y < N; y++) {
                if (board.get(x).get(y).size() == 0) {
                    continue;
                }

                for (Fireball fb : board.get(x).get(y)) {
                    count += fb.m;
                }
            }
        }

        return count;
    }

    static class Fireball {
        int m; // 질랑
        int d; // 방향
        int s; // 속력

        public Fireball(int m, int s, int d) {
            this.m = m;
            this.s = s;
            this.d = d;
        }
    }
}
